### DISCUSSION TITLE
height_of returns same height for every character

### DISCUSSION THREAD

**Discussion by smithtim** [prawnpdf/prawn#1180] [Q&A] [answered]
Whereas `width_of` returns different widths for different characters, `height_of` seems to return the same height for every character.  Is this the expected behavior?  If so, is there a way to get the actual heights?

```ruby
pdf = Prawn::Document.new

puts pdf.width_of 'A'
puts pdf.width_of 'a'

puts pdf.height_of 'A'
puts pdf.height_of 'a'
```

outputs

```
8.004
6.672
13.872
13.872
```

**Comment by gettalong** [comment]
The width of each character has to be different because we need it to calculate the width of a line. However, the height of a line should *not* depend on the characters therein. For example, a line with just `a` should have the same height as a line with just`T`, given they use the same font and font size. Otherwise the lines would not look good/correct. *How* the height is calculated is kinda complicated and may not be the same everywhere.

I don't know if Prawn provides direct access to the font to get the real height but you can always fall back to ttfunk to get the correct values.

**Comment by pointlessone** [comment]
`height_of` returns a height of a bounding box for the text. In case of a single character it's effectively a line height.

As @gettalong mentioned, you can turn to TTFunk for specific character metrics but you won't get exact bounding boxes of glyphs. Most fonts try to keep character metrics height the same to have consistent line height. IF you would like exact bounding boxes of glyps then, unfortunately, I don't believe TTFunk will be of much help as it doesn't parse glyph tables because in Praws we don't use that information for anything so we don't spend time on parsing it.

**Comment by smithtim** [comment]
Thank you for the replies.  Here is my use case: given a character, I want to center it in a box.  Just one character, not a line of text.  I tried `align` and `valign`:

```ruby
pdf = Prawn::Document.new
pdf.stroke_rectangle [0, pdf.margin_box.height], 15, 15
pdf.text_box '.', :width => 15, :height => 15, :align => :center, :valign => :center
```

which produces
![image](https://user-images.githubusercontent.com/25237756/97096329-6806b780-161f-11eb-82ab-cccea5e83c67.png)
The character is centered horizontally in the box, but not vertically.

I thought if I could get the character's width and height, I could do the calculations and center it myself.  But if `height_of` returns the same height for every character, that won't work.

Do you think this centering problem could be solved with TTFunk somehow, or am I out of luck?

**Comment by gettalong** [comment]
I have looked at the code of the font implementation of Prawn:

* The bounding boxes of individual glyphs of the standard 14 PDF fonts (e.g. Helvetica, Times Roman, Courier, ZapfDingsbat and Symbol) are somewhat parsed but not fully. So the information is there. However, it is not available to the user of the library. 

* For TrueType fonts (I guess with the newest version of ttfunk also OpenType fonts but have not checked), the information of the glyph bounding boxes should be available with `font.glyph_outlines.for(glyph_id)`. The returned object has `#x_min`, `#y_min`, `x_max`, and `#y_max` accessors which should give you the needed information. The `font` object is available in Prawn `pdf.font.ttf`

**Comment by smithtim** [comment, marked as answer]
I got it working!  Here is the code (using a font that is available on my system).  Thank you for your help.

```ruby
def center_at pdf, character, x, y
  character_code = character.unpack1 'U*'
  glyph_id = pdf.font.ttf.cmap.unicode.first[character_code]
  outlines = pdf.font.ttf.glyph_outlines.for glyph_id
  units_per_em = pdf.font.ttf.header.units_per_em.to_f

  left = outlines.x_min / units_per_em * pdf.font_size
  right = outlines.x_max / units_per_em * pdf.font_size
  bottom = outlines.y_min / units_per_em * pdf.font_size
  top = outlines.y_max / units_per_em * pdf.font_size

  [x-(left+right)/2.0, y-(top+bottom)/2.0]
end

pdf = Prawn::Document.new
pdf.font "/run/current-system/sw/share/X11-fonts/DejaVuSans.ttf"

pdf.stroke_rectangle [0, pdf.margin_box.height], 15, 15

character = '.'
center_x = 15/2.0
center_y = pdf.margin_box.height - 15/2.0
x, y = center_at pdf, character, center_x, center_y
pdf.draw_text character, :at => [x,y]

pdf.render_file 'out.pdf'
```

which produces
![image](https://user-images.githubusercontent.com/25237756/97773480-45880900-1b0d-11eb-91cb-fa76f3056d47.png)
And here is a solution to the original problem of finding the height:

```ruby
def dimensions pdf, character
  character_code = character.unpack1 'U*'
  glyph_id = pdf.font.ttf.cmap.unicode.first[character_code]
  outlines = pdf.font.ttf.glyph_outlines.for glyph_id
  units_per_em = pdf.font.ttf.header.units_per_em.to_f

  width = (outlines.x_max-outlines.x_min) / units_per_em * pdf.font_size
  height = (outlines.y_max-outlines.y_min) / units_per_em * pdf.font_size

  [width, height]
end

pdf = Prawn::Document.new
pdf.font "/run/current-system/sw/share/X11-fonts/DejaVuSans.ttf"

width, height = dimensions pdf, 'A'
puts "width = #{width}, height = #{height}"

width, height = dimensions pdf, 'a'
puts "width = #{width}, height = #{height}"
```

outputs

```
width = 8.015625, height = 8.748046875
width = 5.54296875, height = 6.890625
```
